devtools/working-with-dsproxy/working-with-dsproxy.md

```diff
@@ -0,0 +1,437 @@
+
+```text
+TubLike tub = TubLike(_tub);
+UniswapExchangeLike daiEx = UniswapExchangeLike(_DAIExchange);
+UniswapExchangeLike mkrEx = UniswapExchangeLike(_MKRExchange);
+
+TokenLike dai = tub.sai();
+TokenLike mkr = tub.gov();
+PepLike pep =   tub.pep();
+```
+
+Convert the input `cupid` into a left-padded bytes32 hex value format that the Tub contract expects.
+
+```text
+bytes32 cup = bytes32(cupid);
+```
+
+#### Set all allowances
+
+Add a `setAllowance` private function to the contract to set allowances
+
+```text
+function setAllowance(TokenLike token_, address spender_) private {
+        if (token_.allowance(address(this), spender_) != uint(-1)) {
+            token_.approve(spender_, uint(-1));
+        }
+    }
+```
+
+In the `wipeWithDai` function, we can now set the required allowances using the `setAllowance` function.
+
+* Allow the Tub contract to debit DAI from the DSProxy contract
+* Allow the Tub contract to debit MKR from the DSProxy contract
+* Allow the Uniswap DAI Exchange contract to debit DAI from the DSProxy contract
+
+```text
+setAllowance(dai, _tub);
+setAllowance(mkr, _tub);
+setAllowance(dai, _DAIExchange);
+```
+
+#### Transfer Dai to the DSProxy contract
+
+Read the current MKRUSD price
+
+```text
+(bytes32 val, bool ok) = pep.peek();
+```
+
+Calculate the stability fees owed for the `wad` amount of dai debt being wiped from the CDP
+
+```text
+uint daiFee = rmul(wad, rdiv(tub.rap(cup), tub.tab(cup)));
+```
+
+Calculate the amount of MKR needed for successfully executing wipe by dividing the stability fee amount accrued in Dai with the current value reported by the MKRUSD price oracle contract
+
+```text
+uint mkrFee = wdiv(rmul(wad, rdiv(tub.rap(cup), tub.tab(cup))), uint(val));
+```
+
+Calculate the additional Dai needed to buy MKR from Uniswap. This is done by first calculating the amount of ETH needed to buy the required MKR, and then the amount of Dai needed to buy the required ETH
+
+```text
+uint ethAmt = mkrEx.getEthToTokenOutputPrice(mkrFee);
+uint daiAmt = daiEx.getTokenToEthOutputPrice(ethAmt);
+```
+
+We can now calculate the total amount of Dai and transfer it from the user's address to their DSProxy contract
+
+```text
+daiAmt = add(wad, daiAmt);
+require(dai.transferFrom(msg.sender, address(this), daiAmt));
+```
+
+#### Exchange Dai for MKR on Uniswap
+
+The `tokenToTokenSwapOutput` function exchanges Dai for the required amount of MKR specified in the input as `mkrFee`. The remaining inputs set the maximum amount of DAI and ETH used for the transaction, and deadline set for the transaction to be valid. Paying stability fees can be skipped if the MKRUSD oracles are not valid.
+
+```text
+if(ok && val != 0) {
+    daiEx.tokenToTokenSwapOutput(mkrFee, daiAmt, uint(999000000000000000000), uint(1645118771), address(mkr));
+}
+```
+
+#### Wipe Dai debt from the CDP
+
+Wipe debt of the CDP with DAI and pay the stability fee with MKR available on the DSProxy contract.
+
+```text
+tub.wipe(cup, wad);
+```
+
```

Comment:
This could be a good point share a link to or a section of the entire `wipeWithDai( )` function or `WipeProxy.sol`